### PR TITLE
Scala Promise and Java CompletableFuture instrumentation

### PR DESCRIPTION
**Pull request by bantonsson**
Instrumentation and tests for Scala Promise/Future and Java CompletableFuture.

### CODE REVIEW

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/DatadogClassLoader.java [hunk 1]
```diff
@@ -67,11 +67,16 @@ public boolean hasLoadedClass(final String className) {
   }
 
   Class<?> loadFromPackage(String packageName, String name) throws ClassNotFoundException {
-    Object packageLock = internalJarURLHandler.getPackageLock(packageName);
+    InternalJarURLHandler.Lock packageLock = internalJarURLHandler.getPackageLock(packageName);
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/DatadogClassLoader.java:70]: There _was_ a memory saving in making these changes, but not a significant improvement in the time to load our classes. I think there is also potential for deadlock. If this is causing problems, I suggest either synchronizing on `getClassLoadingLock(name)`, or completely abandoning the package based classloading logic.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      ContextStore<K, ConcurrentState> contextStore, K key) {
+    final ConcurrentState state = contextStore.get(key);
+    if (state == null) {
+      return;
+    }
+    state.closeAndClearContinuation();
+  }
+
+  boolean setContinuation(TraceScope.Continuation continuation) {
+    final boolean result = continuationRef.compareAndSet(null, continuation);
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java:78]: I think the changes made to `State` were an improvement here, and the assert I added previously was incorrect.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    return blockedClasses.get(type.ordinal()).contains(className);
+  }
+
+  private static final ClassValue<Integer> SKIP =
+      new ClassValue<Integer>() {
+        @Override
+        protected Integer computeValue(Class<?> clazz) {
+          int result = 0;
+          String name = clazz.getName();
+          for (SkipType type : SkipType.values()) {
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java:58]: Pet peeve - `SkipType.values()` allocates every time. I'd prefer an `EnumSet` or `SkipType[] VALUES = values()` inside the enum.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  public static boolean skip(SkipType type, String className) {
+    return blockedClasses.get(type.ordinal()).contains(className);
+  }
+
+  private static final ClassValue<Integer> SKIP =
+      new ClassValue<Integer>() {
+        @Override
+        protected Integer computeValue(Class<?> clazz) {
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java:55]: Isn't this mask basically an `EnumSet<SkipType>`?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java (hunk 3 above)
Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java:58]: Ah, didn't know that.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      return mask;
+    }
+
+    SkipType() {
+      this.mask = 1 << this.ordinal();
+    }
+  }
+
+  public static boolean skip(SkipType type, Object instance) {
+    return (SKIP.get(instance.getClass()) & type.getMask()) != 0;
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java:45]: Could be `EnumSet.contains(type)` which reduces to bitwise operations too.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+          for (SkipType type : SkipType.values()) {
+            if (skip(type, name)) {
+              result |= type.getMask();
+            }
+          }
+          return result;
+        }
+      };
+
+  private static final ArrayList<Set<String>> blockedClasses =
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java:67]: This could be an `EnumMap<SkipType, Set<String>`

File: dd-java-agent/instrumentation/guava-10/src/test/groovy/ListenableFutureTest.groovy [hunk 7]
```diff
@@ -32,7 +32,7 @@ class ListenableFutureTest extends AbstractPromiseTest<SettableFuture<Boolean>,
   }
 
   @Override
-  Boolean get(SettableFuture<Boolean> promise) {
+  boolean get(SettableFuture<Boolean> promise) {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/guava-10/src/test/groovy/ListenableFutureTest.groovy:35]: 🥇

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java (hunk 5 above)
Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/SkipFilter.java:45]: Yes, you are completely correct. I'll change it to `EnumSet` and `EnumMap`

File: dd-java-agent/instrumentation/java-concurrent/java-completablefuture/src/main/java8/java/util/concurrent/CompletableFutureAdvice.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+import datadog.trace.context.TraceScope;
+import java.util.concurrent.CompletableFuture.UniCompletion;
+import net.bytebuddy.asm.Advice;
+
+// This class is put into java.util.concurrent to allow access to package private classes.
+public final class CompletableFutureAdvice {
+
+  public static final class UniConstructor {
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void afterInit(@Advice.This UniCompletion thiz) {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/java-concurrent/java-completablefuture/src/main/java8/java/util/concurrent/CompletableFutureAdvice.java:18]:
```suggestion
    public static void afterInit(@Advice.This UniCompletion zis) {
```

File: dd-java-agent/instrumentation/java-concurrent/java-completablefuture/src/test/groovy/CompletableFutureTest.groovy [hunk 9]
```diff
@@ -178,26 +232,47 @@ class CompletableFutureTest extends AgentTestRunner {
         basicSpan(it, "child", span(0))
       }
     }
+
+    where:
+    value << (0..((1 << 3) - 1))
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/java-concurrent/java-completablefuture/src/test/groovy/CompletableFutureTest.groovy:237]:
```suggestion
    value << (0..7)
```

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/DatadogClassLoader.java (hunk 1 above)
Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/DatadogClassLoader.java:70]: I haven't seen the locks causing any issues.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java (hunk 2 above)
Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java:78]: Yeah, I forgot to remodel my things in the same way. I'll take a look at it.

File: dd-java-agent/instrumentation/java-concurrent/java-completablefuture/src/main/java8/java/util/concurrent/CompletableFutureAdvice.java (hunk 8 above)
Comment (bantonsson) [reply dd-java-agent/instrumentation/java-concurrent/java-completablefuture/src/main/java8/java/util/concurrent/CompletableFutureAdvice.java:18]: Yes, that is easier to distinguish.

File: dd-java-agent/instrumentation/java-concurrent/java-completablefuture/src/test/groovy/CompletableFutureTest.groovy (hunk 9 above)
Comment (bantonsson) [reply dd-java-agent/instrumentation/java-concurrent/java-completablefuture/src/test/groovy/CompletableFutureTest.groovy:237]: Yes, that is a bit convoluted. I'll change it and add a comment.

**Review by richardstartin — APPROVED** [review]
I would like to get this merged quickly to integrate more changes with it

**Review by tylerbenson — COMMENTED** [review]
Oops... I forgot to submit my review last week.  Adding these comments even though they're likely out of date.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java [hunk 10]
```diff
@@ -95,8 +96,23 @@ public static ResettableClassFileTransformer installBytebuddyAgent(
       agentBuilder = agentBuilder.with(listener);
     }
     int numInstrumenters = 0;
-    for (final Instrumenter instrumenter :
-        ServiceLoader.load(Instrumenter.class, AgentInstaller.class.getClassLoader())) {
+    ServiceLoader<Instrumenter> loader =
+        ServiceLoader.load(Instrumenter.class, AgentInstaller.class.getClassLoader());
+    for (final Instrumenter instrumenter : loader) {
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java:101]: Could you add a comment on why this should be a separate iteration loop.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
+package datadog.trace.bootstrap.instrumentation.java.concurrent;
+
+import datadog.trace.bootstrap.ContextStore;
+import datadog.trace.context.TraceScope;
+import java.util.concurrent.atomic.AtomicReference;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public final class ConcurrentState {
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java:9]: Why is this needed in addition to `State`?

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    final ConcurrentState state = contextStore.get(key);
+    if (state == null) {
+      return null;
+    }
+    return state.activateAndContinueContinuation();
+  }
+
+  public static <K> void closeScope(
+      ContextStore<K, ConcurrentState> contextStore, K key, TraceScope scope, Throwable throwable) {
+    final ConcurrentState state = contextStore.get(key);
```
Comment (richardstartin) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java:63]: If there's no scope we should check that first, in case `ContextStore` is backed by something expensive

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ExcludeFilterProvider.java [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+import datadog.trace.bootstrap.instrumentation.java.concurrent.ExcludeFilter.ExcludeType;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Used to allow an {@link Instrumenter} to opt out of broad instrumentations like {@link Runnable}.
+ *
+ * <p>These are looked up in a separate pass before the {@link Instrumenter} is allowed to add
+ * instrumentations. Note, it is up to the {@link ExcludeFilterProvider} to check if it is enabled
+ * and only to
```
[the remark this answers is not in the document]
Comment (bantonsson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/ExcludeFilterProvider.java:12]: Yeah. Only to...

Thanks. I'll update the comment.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ExcludeFilter.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
+package datadog.trace.bootstrap.instrumentation.java.concurrent;
+
+import java.util.EnumMap;
+import java.util.EnumSet;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class ExcludeFilter {
```
[the remark this answers is not in the document]
Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ExcludeFilter.java:9]: So it is needed because the normal `Runnable` and `Executor` instrumentations get in the way, and do double or triple work, including capturing contexts and adding fields that should never be used, to the `UniCompletion` instances.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java (hunk 10 above)
Comment (bantonsson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/AgentInstaller.java:101]: Sure.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java (hunk 11 above)
Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java:9]: It's because it has different semantics, than `State` since there is a benign race between multiple threads in the `tryFire` method, that I thought was better to keep separate from the normal `State` instead of adding other methods onto `State` that might be used inadvertently.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java (hunk 12 above)
Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java:63]: Ok, I'll look over the code for this pattern.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/DatadogClassLoader.java [hunk 15]
```diff
@@ -159,11 +164,16 @@ public DelegateClassLoader(
     @Override
     protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
       String packageName = shared.getPackageName(name);
-      Object packageLock = internalJarURLHandler.getPackageLock(packageName);
+      InternalJarURLHandler.Lock packageLock = internalJarURLHandler.getPackageLock(packageName);
```
[the remark this answers is not in the document]
Comment (bantonsson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/DatadogClassLoader.java:167]: Sure. I'll do that.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java (hunk 12 above)
Comment (tylerbenson) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ConcurrentState.java:63]: > Ok, I'll look over the code for this pattern.

The mark of a good senior engineer.  👍

**Review by devinsba — APPROVED** [review]
I can't find any reason this shouldn't be merged after the clarifying comments are added. It's a big change though so maybe we should wait for a 3rd 👍

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ExcludeFilter.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+        case "java.lang.Runnable":
+          return RUNNABLE;
+        case "java.util.concurrent.Callable":
+          return CALLABLE;
+        case "java.util.concurrent.Future":
+          return FUTURE;
+        case "java.util.concurrent.ForkJoinTask":
+          return FORK_JOIN_TASK;
+        default:
+          return null;
```
Comment (BurgersMcSlopshot) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ExcludeFilter.java:33]: I know this is quite a bit late to the game, but returning null here as an enum value especially with no documentation in the code as to why seems like a footgun.  Would it be better to throw instead?

Comment (BurgersMcSlopshot) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/java/concurrent/ExcludeFilter.java:33]: returning null here is very weird and seems like it could be particularly troublesome in the future.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/FieldBackedProvider.java [hunk 17]
```diff
@@ -461,13 +453,27 @@ public static void resetContextMatchers() {
...
       @Override
       public boolean matches(
           final TypeDescription typeDescription,
           final ClassLoader classLoader,
           final JavaModule module,
           final Class<?> classBeingRedefined,
           final ProtectionDomain protectionDomain) {
+
+        // First check if we should skip injecting the field based on the key type
+        if (skipType != null && ExcludeFilter.exclude(skipType, typeDescription.getName())) {
```
Comment (BurgersMcSlopshot) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/FieldBackedProvider.java:467]: to pair with the comment in the ExcludeFilter regarding potentially coercing a string to a null enum value, why not pass keyType directly into the filter and skip the null check here?

Comment (bantonsson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/context/FieldBackedProvider.java:467]: Thanks. That is a very good point.
